### ISSUE TITLE
[Bug]: GlideN64 Crashes but Jabo's works fine

### ISSUE THREAD

**Issue by cursedUs64-git** [project64/project64#2361]
### Expected behavior?

i thought it would open as normal.

### Actual behavior?

![Screenshot (3)](https://user-images.githubusercontent.com/110907873/234966193-791a9527-56a1-49df-be48-ad940c0f63c0.png)

### Steps to reproduce the problem

1. have an old pc
2. install latest version of project64
3. any rom you play gives this error message
4.
5.
...

### Error messages?

its in the above picture

### What game were you playing?

Title: SAE 64 BE

Nintendo 64 game or Nintendo 64 Disk Drive: Nintendo 64 game

ROM information (First you must uncheck "hide advanced settings" in the settings menu, then File > ROM info should show up, post the information contained there in this field. Alternatively, you can right click the ROM in the ROM browser and hit "ROM information".):
![Screenshot (51)](https://user-images.githubusercontent.com/110907873/235981519-f04f94c1-07d1-4b85-abe2-c3aa5866c324.png)

Save type: (attach game save or save state here)

Original or modified ROM?:modified rom, wont give patch because its WIP.

### Specifications and build details

Project64: Dev-4.0.0-6108-5cfb80f x64

OS: (include version and OS build number using WinVer): windows 10 22h2 19045.2130

CPU: Intel Core 2 Duo e7200

RAM: 4GB RAM

Graphics: Integrated Graphics (intel g33/g31 express chipset family (microsoft corporation - wddm 1.0)

### Using included plugins

I can confirm I am using the included RSP, audio, input, and graphics plugins

### Plugins used while the issue occurred

Graphics plugin (and HLE or LLE): GlideN64 HLE
Audio plugin: project64 audio plugin
Input plugin: project64 input plugin
RSP plugin: rsp plugin dev

### Copyright Compliance

- [X] There are no infringing files attached to this bug report.

**Comment by DerekTurtleRoe** [comment]
@cursedUs64-git Please fill out the rest of the bug report

**Comment by flagrama** [comment]
Can you clarify whether this is a custom build of the game, or an original dump? It looks like you created a custom build in WSL and are trying to run it, and if it is a custom build could you try with an original dump because there could be issues with your toolchain or code causing the issue.

**Comment by Squall-Leonhart** [comment]
lmao

closed in 3,2, 1

**Comment by cursedUs64-git** [comment]
I am using an original dump, i opened both, custom and original but they both dont work.

**Comment by flagrama** [comment]
Write that in the edit to your bug report filling out all the other information you skipped, not in a comment. It needs to be clarified as part of the bug report.

**Comment by Squall-Leonhart** [comment]
User error.

![image](https://user-images.githubusercontent.com/1098176/235139312-031337b1-6bb1-48db-99c2-fdb90856efdf.png)
![image](https://user-images.githubusercontent.com/1098176/235140396-7a6ee92d-9548-4010-9393-e8fff6045a6b.png)

**Comment by cursedUs64-git** [comment]
Hey, i updated my comment. Now i use a modified rom but it still says that error.

**Comment by Squall-Leonhart** [comment]
There is little to no opengl support on that ancient intel gpu, GlideN64 requires a a GPU supporting opengl 4.1 at the very least (or 4.3), either way, your igp has neither.

**Comment by cursedUs64-git** [comment]
> There is little to no opengl support on that ancient intel gpu, GlideN64 requires a a GPU supporting opengl 4.1 at the very least (or 4.3), either way, your igp has neither.

so what do i do

**Comment by DerekTurtleRoe** [comment]
> > There is little to no OpenGL support on that ancient Intel GPU, GLideN64 requires a a GPU supporting OpenGL 4.1 at the very least (or 4.3), either way, your iGPU has neither.
> 
> so what do i do

Well, first, as Squall said, your system is old enough that it doesn't support any modern graphics API that GLideN64 requires. But also, accurate N64 emulation takes a lot of horsepower to run, and this system just doesn't have it.

Another issue is this is a ROM hack, so that can cause its own issues as well. If this also occurs on vanilla Super Mario 64 then it might've been worth investigating, but we would then ask for the patch, so we can look into it. And as you've said, it is a "WIP" so my thoughts would be to wait until it is either publically available or wait until it is finished to open issues about it in emulators.

Lastly, I think testing this ROM hack on other emulators and plugins might be a good idea for lower end hardware, so something like Project64-video, Glide64, etc. Either that or try Rosalie's Mupen GUI and see if you can get it working there.

**Comment by flagrama** [comment]
I remember having an Intel GMA like that back in 2007. Those are so terrible. Nothing works right on it because Intel didn't care to have decent OpenGL support until the iGPUs they made around like 2012. You can probably try to install some [ancient drivers from the Intel site](https://www.intel.com/content/www/us/en/support/products/81519/graphics/legacy-graphics/graphics-drivers-for-intel-g33-express-chipset.html) to see if it works any better since the ones installed by Windows Update straight up have inferior OpenGL support, but that usually only works if you're having issues with Intel HD iGPUs and nothing can really help a GMA.

**Comment by dftf-stu** [comment]
@cursedUs64-git: When you first open Project64 on that PC, choose the option "Project64 Video (for older PC)".  (Or on your current download of the app, go to Options > Configuration > Plugins and then change the "Video (graphics) plugin" dropdown to the "Project64 Video" option, click OK, fully-close the app, then bring it back up.)

There is no-guarantee this will work though: the [Intel page](https://www.intel.co.uk/content/www/uk/en/products/sku/35348/intel-core2-duo-processor-e7200-3m-cache-2-53-ghz-1066-mhz-fsb/specifications.html) for your processor says it came-out around the middle of 2008.  While it should support OpenGL 2.0 (which is what the "Project64 Video" plugin requires, I've been told previously), the drivers for your CPU and GPU will be so-old Windows 10 is unlikely to actually be able to use that feature.  Instead Windows 10 will "emulate" OpenGL 2.0, which means (1) games will run at a lower-speed and (2) you cannot use Project64 in full-screen mode.

If the "Project64 Video (for older PC)" option doesn't work, then you have two options:
- buy a newer computer
- run Windows 7 on your computer, where OpenGL 2.0 should be fully-supported.  However remember that the Project64 app itself is unlikely to work on Windows 7 forever, so long-term buying a newer PC would be your better-option

**Comment by flagrama** [comment]
That's nonsense? First, GMAs are [part of the chipset](https://ark.intel.com/content/www/us/en/ark/products/31914/intel-82g33-graphics-and-memory-controller.html), not the CPU. Intel's iGPUs didn't end up as part of the CPU until the core-i series stuff pretty much. The Intel hardware/drivers for these iGPUs have never and will never implement all aspects of OpenGL. They don't even say that it supports OpenGL on ARK. Windows 7 is no different than Windows 10. It sounds like you're talking about GLSL compatibility profiles, which again, depend on graphics hardware and driver support, the OS doesn't have a bearing on that. Intel just straight up did not implement certain OpenGL features, not even for the compatibility layer. Windows 7 will make no difference.

**Comment by Squall-Leonhart** [comment]
I think they are talking about the fact that the windows 10 compatible drivers for these older chipset gpu's are provided via windows update and have never been provided by intel in an installable package.

**Comment by dftf-stu** [comment]
@flagrama: I'm not a developer or coder so apologies if I'm using incorrect terms in areas.

But essentially one of my old laptops (from around 2010) has a first-generation Intel Core i3 processor ([i3-330M](https://www.intel.co.uk/content/www/uk/en/products/sku/47663/intel-core-i3330m-processor-3m-cache-2-13-ghz/specifications.html)), with integrated "Intel HD" graphics.  An old Intel page (which I'm struggling to find now, but was similar to [this one](https://www.intel.com/content/www/us/en/support/articles/000005526/graphics.html)) says OpenGL 2.0 hardware-acceleration is not supported by their drivers for first-gen "Core" products in anything newer-than Windows 8.0, or 8.1 for second-gen "Core" products.  Meaning if you run Windows 10 on such a device, OpenGL 2.0 is essentially being "emulated" by Windows, similar to how old DirectX games (such as DirectX 9) run via a translation-layer on Intel Arc GPUs (look up "D3D9On12").

I can run Project64 under Windows 10 on this laptop, but often only in windowed-mode: switching to full-screen either crashes the app, or makes it run much-more slowly, with flickering.  However, if I take the hard-drive out, and pop-in an old hard-drive which has a Windows 7 install I used to use on this laptop, Project64 (including the latest Dev builds) run-faster, including in full-screen without any issue.  I'm guessing for the OP, their GPU or iGPU or whatever simply doesn't support all the features under Windows 10 as it once did under 7 or Vista that the device originally shipped-with.

(Now ironically I can run PS1 and Saturn games at full-speed on that laptop, and can even run PS2 games at near full-speed (obviously tweaking settings to their lowest levels), mostly because PCSX2 actively supports DirectX, which my iGPU does still support at a hardware-level under Windows 10.  Whereas with Project64, the DirectX support is "essentially dead" and the OpenGL 2.0-supporting Project64 Video plugin "shouldn't really be used".  So I think long-term it would be good for the Project64 dev-team to ask "what CPUs and GPUs should we actually support" as I do find this rather unclear.  If you offer a plug-in labelled "for older PCs", it's not surprising those of us with older PCs then expect it will work.  Maybe you need to set some kind of rules around exactly "how old" such a device can actually be?)

**Comment by dftf-stu** [comment]
@Squall-Leonhart: Intel no-longer offer driver-downloads for some legacy products, yes, as stated [here](https://www.intel.com/content/www/us/en/support/articles/000005526/graphics.html).  However my main-point was that OpenGL 2.0 support at a **hardware-level** isn't supported in newer-versions of Windows for some old GPUs or iGPUs. 

Similar to the issue of newer-versions of Project64 not running on Windows XP (unless you edit the config file to disable the Discord support) or Vista (due to an API change that occurred in Windows 7, but the app expecting this change on Vista), maybe the dev-team need to decide which CPUs, GPUs and iGPUs (along with versions of Windows) this app will actually support?

I mean, I've just played a number of N64 games recently on my first-gen Intel Core i3 (such as Banjo-Kazooie, Banjo-Tooie, Rayman 2 and Tonic Trouble) in windowed-mode using the "Project64 (for older PCs)" video plug-in.  Yes, full-speed isn't maintained at all times (especially in areas with a high polygon-count), but they otherwise run without any graphical issues I could notice.  HOWEVER while most recent Dev versions run without issue (4.0.0-6108, 4.0.0-6111, 4.0.0-6117 and 4.0.0-6120) the most-recent version, 4.0.0.6131, no-longer works: the "percentage of full speed" gets-up to insane values (such as 15000%) and no ROMs load.  So... is it worth me logging a ticket to say there is an issue, or because my iGPU isn't officially supported for OpenGL 2.0 under Windows 10, should I assume I'm unsupported.  This project doesn't really make-clear exactly what it supports and what it doesn't...

**Comment by cursedUs64-git** [comment]
Hey, Stu. You can just post an "issue" with the following you addressed as
a suggestion.

**Comment by dftf-stu** [comment]
@cursedUs64-git: I could log a new issue on here, but as I said: is it worth it?  Are old devices like yours and mine supported by Project64 or not?  There's no-point me logging a ticket if the answer is "no", but the people behind the project really need to be the ones to make a decision on that.

Currently the "Project64 Video" plug-in is stated as being "for older PCs", but perhaps they need to qualify _how-old_ they mean.  Maybe third-gen Intel Core should be the floor?  And for AMD... not-sure, but [this](https://www.amd.com/en/support/kb/faq/gpu-615) article might help inform a decision?

**Comment by dftf-stu** [comment]
@cursedUs64-git: by-the-way, if you go to [this Intel page](https://www.intel.co.uk/content/www/uk/en/support/articles/000005524/graphics.html) and click on the "Older Graphics Products" heading you will find your G31/G33 listed in the middle of that table:

![Screenshot 2023-05-19 145625](https://github.com/project64/project64/assets/83965343/a7a822ad-1a6d-4b45-a57e-a7c8b77adf99)

Your GPU only-supports OpenGL 1.4 at-most, meaning it won't-work with the "Project64 Video" plug-in, as that requires at-least OpenGL 2.0.  You could try the "Jabo" DirectX plug-in, as that requires DirectX 8 or above, and your GPU supports DirectX 9 so it may work -- BUT even if it does, expect a lot of visual-glitches when using it, as it hasn't been supported for years.

As your CPU [is listed](https://www.intel.co.uk/content/www/uk/en/products/sku/35348/intel-core2-duo-processor-e7200-3m-cache-2-53-ghz-1066-mhz-fsb/specifications.html) as being a desktop one, then assuming that is correct and you do have a desktop PC, you could look to upgrade the graphics-card to one that supports at-least OpenGL 2.0 (or even OpenGL 3.3)?

**Comment by flagrama** [comment]
> maybe the dev-team need to decide which CPUs, GPUs and iGPUs (along with versions of Windows) this app will actually support?

From the readme

> Graphics card
> - DirectX 8 capable (Jabo's Direct3D8)
> - OpenGL 2.0 capable (3.0+ recommended) (Project64 Video)
> - OpenGL 3.3 capable (GLideN64)
> 
> Intel integrated graphics can have issues that are not present with Nvidia and AMD GPU's even when the requirements are met.

**Comment by cursedUs64-git** [comment]
I really can't buy a new graphics card since they are too expensive in my
country so ig i have to stick with jabo's

**Comment by dftf-stu** [comment]
@flagrama: yes, the [README](https://github.com/project64/project64#readme) there does indeed offer some guidance.

But there are some-issues with that:
1. You have to scroll-down that page to actually come-across it, for-starters; I'd bet many-users have never once seen it!
2. But for those who have, how-many will realistically know what it means?  Many-users find it difficult-enough to even know which Windows is on their-device, let-alone know what "OpenGL" means, or what version of it their GPU or iGPU can support!  (And AFAIK, there isn't any easy-way within Windows to find this out: you'd have to use a third-party app like [GLView](https://realtech-vr.com/home/glview))
3. Many-users looking to download Project64 will search for it, and click the link to the official-website¹ (https://www.pj64-emu.com).  That site doesn't-mention anywhere any hardware-requirements, or which versions of Windows are supported, nor does it say to go to the README page on here for that
4. Even assuming a user does end-up on Github, does find the README and does understand the requirements, it still doesn't answer whether "emulated OpenGL 2.0" is supported by the legacy "Project64 Video" plug-in or not
5. Where Windows support is concerned, those minimum-requirements are incorrect: even assuming they apply to only the current stable-release, not the Dev releases (which is not made-clear), the latest stable-releases do not run on XP (without disabling Discord by adding the "Enable Discord RPC=0" line into the Project64.cfg file manually), despite it being listed.  Vista isn't mentioned there, but support could be implied by "limited support for Windows 8.1 and below".

¹In Google for example: searching for "Project64" the official-website is the first-result; second is the Wikipedia article for it; third is a site called "Emulator Zone"; and this GitHub project for it is the fourth-result.

**Comment by Squall-Leonhart** [comment]
In actuality, the PJ64 Video plugin requires atleast opengl 3.3 for general function, anything lower is not going to completely emulate the various microcodes that require FBE's, and there is even no guarantee of NPOT compatibility from the driver on 2.0

**Comment by dftf-stu** [comment]
@Squall-Leonhart: apologies but as I've said before I'm not a developer or coder so I've no idea what "FBE" or "NPOT" mean.  I assume they are some instructions or features the N64 hardware offered, but that would be a guess.

But as an end-user the [README](https://github.com/project64/project64#readme) says "OpenGL 2.0 capable (3.0+ recommended) (Project64 Video)" so that's what realistically I can be expected to go-by.  If certain-games have visual-effects that won't work on that plug-in when anything-less than OpenGL 3.3 is offered, then a compatibility-list somewhere would be the useful-place to list that, on a per-game basis, but nothing like-that exists thesedays, and Project64 no-longer ships with the "Notes" columns visible in the ROM Browser by-default either (though I've read-before the information they would present is massively out-of-date anyway, last-updated during the v1.6 era of this app).

**Comment by DerekTurtleRoe** [comment]
@dftf-stu Some of those points are valid, and I will update the README with some fixes tonight, but at the same time we are a small team and it is mostly Zilmar doing the work and the rest of us just test things, make small fixes, and report bugs. As much as we'd all like to maintain support for any system to be able to play N64 games at full speed, at some point we just have to adjust the requirements because we don't realistically have the manpower and time to support such a massive undertaking supporting multiple legacy versions of Windows. Windows 8.1 and below are all not supported by Microsoft anymore, and most of those operating systems/hardware don't get driver updates either. Older versions of Windows 10 are also not going to be supported soon, so at some point we have to ask ourselves if it is worth it supporting older versions of Windows in comparison to just supporting things we know work.

Also the emulator is in a state of flux right now being in the middle of some massive changes, so things are going to be broken for a bit (it is for the greater good).

For your information, FBE is frame buffer emulation, which when disabled can be a nice speed improvement, but many games don't run without it.

NPOT is non-power-of-two I believe, which is a type of texture that does not have dimensions that are a power of two in size. Basically it's dimemsions aren't divisible by two.

Also side note, most of the things in the "notes" column have been removed since they weren't needed, and I think that column will be removed at some point.

**Comment by dftf-stu** [comment]
@DerekTurtleRoe: I appreciate you're a small-team, and Project64 is great for what it is.  But sometimes some-focus needs to go on the little-things.  For example, if you un-tick the "Automatically compress saved states" option, then this breaks saved states entirely: when you go to load one, you get "Fatal error: stopping emulation".  Squall replied in a previous report (#2260) I made about this in August last-year that this is a known-issue, and the option needs to be removed from the UI.  But yet, in the current Dev builds, it's still there by-default.  Why not remove that option sooner, to avoid users accidentally breaking the app -- such as the user who reported what might be the same issue (#2324)?

As for Windows support, I entirely agree this project should not support old versions of Windows forever.  All I'm saying is you need to have some-sort of roadmap or plan as to when you will drop-support, and publicly communicate that.  As of now, Windows XP can still run the latest stable and Dev releases -- but only if the user knows (as it's not documented anywhere, nor does the app ship with any help files) that they have to turn-off Discord support by manually-editing the "Project64.cfg" file, or the app won't open.  On Vista, the recent versions of the app don't work full-stop (#1429, #2253).  So... by the final, stable release of the 4.0 version, someone needs to decide if both issues will be addressed, or whether to officially cease support.

Long-term, some sort of help documentation (such as a Wiki, or even just a page on here) would be great too.  If you right-click on a ROM and go into "Edit game settings", it would be nice to know what some of the stuff there (e.g. "Delay DP Interrupt" or "Advanced block linking") actually do, and whether any could provide a speed-hack.  But there isn't any official help anywhere; the best users can do is search Google and see if they can find it mentioned on a random website that still exists, and still loads.

And then there are basic missing features like the ability to map more than one button to a control (for example, being able to setup a keyboard and controller without having to re-map every control each-time you switch between-them).  And the default "Project64 Input Plugin" doesn't offer a "turbo" feature; you have to use the "N-Rage" one, but every single set of instructions I found (mostly on old websites that struggle to load) for how you set this up, nothing would work.

In short: most of the time I have an issue with Project64, it's not because a game won't run, or there are graphical issues.  It's because of a lack of help or documentation; a lack of options within the default plug-ins to do what I need; old features still being in the app, even-though they now crash it, or do nothing (the "Notes" columns in the ROM Browser), or that people just assume everyone using Project64 has "been here since the start" and so must know everything anyway.

**Comment by DerekTurtleRoe** [comment]
@dftf-stu Save states are going to be very hit or miss because we are making a ton of core changes, and a save state is the current CPU state, and memory state saved into a file. So while we are working on the CPU in general and especially the recompiler, it might get broken. I'll see if it is a simple fix, but my guess would be it might be a bit more involved.

Better documentation is something I'm always working on, and actually an explanation of those options is available in the repo, but not in Markdown for easy viewing, so I'll work on that. However as you previously stated, users have a hard time reading sometimes, so I think it won't get read very often. But it would be nice to have it in a more readable format (although a txt file is hardly something you can't view in a browser, but Markdown means one less button to press, so...)

We know there are some commonly suggested features in regards to input, but until we are done porting to other platforms it doesn't make sense to put a ton of effort into that right now, since we'd have to just rewrite it later anyway.

Input plugins I can't comment much on since they aren't maintained by us, except Project64-input, but that is meant to be a simple and easy to use input plugin. I'll start making a list of things commonly requested, but there is no guarantee we will have time/resources to implement any of it.

Discord-rpc I think has been [discontinued](https://github.com/discord/discord-rpc), so we may have to just remove that.

**Comment by cursedUs64-git** [comment]
Hey guys, should i close this issue? I found gln ver 0.3 and it supports opengl 1.4

**Comment by dftf-stu** [comment]
@DerekTurtleRoe: thanks for the reply, but yes, probably best to continue the discussion elsewhere.  But the issue is: where?  I don't really want to use Discord, as it gets tiring thesedays having to join every website out there just to discuss things with projects or companies.  But the "Discussions" area on this website seems forgotten about: many of the discussions there have zero replies, so I'd feel I'd be wasting my time bothering to post there if no-one will ever see it.

@cursedUs64-git: yes, I'd suggest you close this if you've found an older plug-in that works for you.

**Comment by DerekTurtleRoe** [comment]
@cursedUs64-git That is what I would recommend, yes. Sorry we couldn't fix the issue for you with other plugins.

@dftf-stu I can understand what you mean. I usually look at it every once in a while, but my computer wasn't working for the last couple weeks so I haven't had the chance. Discussions are kind of not used properly by users as well as poorly documented by GitHub, but it seems to work OK mostly.

If you want we can discuss things there, although I would say as per the usual just have some patience since people have responsibilities and such and not everybody is on there 24/7.

**Comment by Squall-Leonhart** [comment]
i still think that option should be removed, theres no reason not to pack states.
